Implementation of the current state of MPI Continuations proposal [WIP]

devreal opened the pull request:
This PR adds an implementation of the MPI Continuations proposal as extension to Open MPI. This proposal is current under discussion in the MPI hybrid and accelerator working group. By integrating it as an extension into Open MPI, I am hoping to provide an up-to-date implementation for the community to experiment with. It reflects the current state of the proposal and will be kept in sync with the evolving proposal.

The implementation is mostly confined to the extension itself, with the exception of hooks in request test and wait functions used to allow polling on a continuation request to complete outstanding continuations. This is needed for the implementation of the `"mpi_continue_poll_only"` info key (see the description below).

*Note*: this PR is currently WIP as it relies on a workaround to define `OMPI_HAVE_MPI_EXT_CONTINUE`, which is used to disable the hooks in request test/wait functions if the extension is not enabled. The underlying issue is that the extensions integration currently does not support including the `mpiext.h` header in implementation files (needed to disable the hooks mentioned above).

### Overview of MPI Continuations

Continuations provide a mechanism for attaching callbacks to outstanding operation requests. A call to `MPIX_Continue` takes a request, a function pointer, a user-provided data pointer, and a status object (or `MPI_STATUS_IGNORE`), along with a continuation request and *attaches* the continuation to the operation request:

```C
MPI_Request req;
// status object has to remain valid until the callback is invoked
MPI_Status *status = malloc(sizeof(MPI_Status));
char *buf = ...;
MPI_Irecv(buf, ..., MPI_ANY_SOURCE, ... &req);
MPIX_Continue(&req, &complete_cb, buf, status, cont_req);
assert(req == MPI_REQUEST_NULL);
```

The ownership of non-persistent requests is returned to MPI and the pointer to the request will be set to `MPI_REQUEST_NULL`. The callback is passed the status pointer and the user-provided data pointer:

```C
void complete_cb(MPI_Status *status, void *user_data) {
  printf("Send completed\n");
  char *buf = (char*)user_data;
  process_msg(buf, status->MPI_SOURCE);
  free(buf); // free the send buffer
  free(status);    // free the status
}
```

The status has to remain valid until the invocation of the callback and is set according to the operation before the callback is invoked.

The continuation is registered with the provided continuation request. The continuation request is a request allocated using `MPIX_Continue_init`:

```C
MPIX_Continue_init(info, &cont_req);
```

Continuation requests may be used to test/wait for completion of all continuations registered with it using `MPI_Test/Wait`. Supported info keys are:

- `"mpi_continue_poll_only"`: if true,  only execute continuations when `MPI_Test/Wait` is called on the continuation request. If false, continuations may be executed at any time inside a call into MPI (inside a callback registered with `opal_progress` in the implementation; default: false).
- `"mpi_continue_enqueue_complete"`: if false, the continuation is executed immediately if the operations are already complete when MPIX_Continue is called. Execution is deferred otherwise (default: false)
- `"mpi_continue_max_poll"`: the maximum number of continuations to execute when calling `MPI_Test` on the continuation request (default: -1, meaning unlimited)

A continuation may in turn be attached to a continuation request, in which case it will be executed once all continuations registered with the continuation request have completed.

In addition to `MPIX_Continue`, the proposal also includes `MPIX_Continueall` which attaches a continuation to a set of requests such that the continuation is executed once all operations have completed.


Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>
Signed-off-by: George Bosilca <bosilca@icl.utk.edu>

devreal committed:
Implementation of the current state of MPI Continuations proposal

Continuations provide a mechanism for attaching callbacks to outstanding
operation requests. A call to MPIX_Continue takes a request, a
function pointer, a user-provided data pointer, and a status object (or
`MPI_STATUS_IGNORE`), along with a continution request:

```
MPI_Request req;
// status object has to remain valid until the callback is invoked
MPI_Status *status = malloc(sizeof(MPI_Status));
char *buf = ...;
MPI_Irecv(buf, ..., MPI_ANY_SOURCE, ... &req);
MPIX_Continue(&req, &complete_cb, buf, status, cont_req);
assert(req == MPI_REQUEST_NULL);
```

The ownership of non-persistent requests is returned to MPI and the pointer
to the request will be set to `MPI_REQUEST_NULL`. The callback is passed
the status pointer and the user-provided data pointer:

```
void complete_cb(MPI_Status *status, void *user_data) {
  printf("Send completed\n");
  char *buf = (char*)user_data;
  process_msg(buf, status->MPI_SOURCE);
  free(buf); // free the send buffer
  free(status);    // free the status
}
```

The status has to remain valid until the invocation of the callback
and is set according to the operation before the callback is invoked.

The continuation is *registered* with the provided contination
request. The continuation request is a request allocated with `MPIX_Continue_init`:

```
MPIX_Continue_init(info, &cont_req);
```

Continuation requests may be used to test/wait for completion of all
continuations registered with cont_req using `MPI_Test/Wait`. Supported
info keys are:

- "mpi_continue_poll_only": only execute continuations when MPI_Test/Wait is
  called on the continuation request (default: false)
- "mpi_continue_enqueue_complete": if true, the continuation is executed
  immediately if the operations are already complete when MPIX_Continue is
  called. Execution is deferred otherwise (default: false)
- "mpi_continue_max_poll": the maximum number of continuations to execute
  when calling MPI_Test on the continuation request (default: -1, meaning unlimited)

A continuation may in turn be attached to a continuation request, in which
case the continuation request will be executed once all continuations
registered with the continuation request have completed.

In addition to MPIX_Continue, the proposal also includes MPIX_Continueall
which attaches a continuation to a set of requests such that the
continuation is executed once all operations have completed.

The implementation reflects the current state of the proposal and may
change at any time, following discussions in the MPI Forum. The goal
of this Open MPI extension is to provide a working implementation to
the community to experiment with this API.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>
Signed-off-by: George Bosilca <bosilca@icl.utk.edu>

bosilca reviewed (commented):
What is the impact of the continuations on the latency ? What if we mix normal requests and continuations in the same app ?

devreal committed:
Move parts of ompi_request_cont_data_t into custom request struct

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Rename functions to avoid confusion

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
SQUASHME: Replace opal_show_help instead of fprintf

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Use OMPI_COPY_STATUS to set status

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Add documentation for ompi_continue_register_request_progress

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal commented:
I pushed a bunch of changes. The biggest change is that a continuation request is now derived from `ompi_request_t`, which shrunk `ompi_request_cont_data_t` considerably and cuts down on the pointer chasing. The latter is now only used to link operation requests to the continuation object and the user-provided status object.

devreal commented:
Oh, and I will squash everything down to one commit once we're done...

bosilca reviewed (commented):
Definitively much cleaner.

devreal committed:
Remove re-iteration of continuation requests in test_any

devreal committed:
Enable the continuations extension only if explicitly requested

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Remove lock out of critical path of creating/completing continuations

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix printing of tear-down warning

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix logic error when creating a new continuation

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Use new OMPI_MPIEXT_continue_POST_CONFIG hook

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Request test/wait: prevent continuation requests from being set to inactive

Continuation requests are persistent requests that are always active.
They are implicitly restarted when the first continuation is registered.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Don't enqueue continuations twice for execution if everything is complete at registration

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Poll-only continuations should only be executed by the thread testing/waiting on them

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Allow the waiting thread to execute continuations if not blocked in the sync

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal commented:
@bosilca I pushed a few changes:

1) Now using the new `OMPI_MPIEXT_*_POST_CONFIG` hook. I hope this is how it is intended to be used.
2) I had to make sure that the state of continuation requests is not changed to inactive in test/wait. They are always active and never explicitly (re)started by the user.
3) Continuations that are poll-only (not executed as part of global progress) will now only be executed by the thread testing/waiting for their completion. We don't merge the local list into the global list anymore but instead allow the waiting thread to execute them if not blocked on the sync. Otherwise they will be executed at the end of the wait, once all continuations are eligible for execution. I think this change makes sense as this avoids other threads from being disturbed by poll-only continuations.

bosilca commented on ompi/mpiext/continue/c/continuation.h:
```diff
@@ -0,0 +1,78 @@
+int ompi_continue_progress_request(ompi_request_t *cont_req);
+
+/**
+ * Attach a continuation to a set of operations represented by \c requests.
```
what is this \c ?

bosilca commented on ompi/mpiext/continue/c/continuation.c:
```diff
@@ -0,0 +1,671 @@
+
+/**
+ * Register the provided continuation request to be included in the
+ * global progress loop (used while a thread is waiting for the contnuation
```
typo. `contnuation`

bosilca commented on ompi/mpiext/continue/c/continuation.c:
```diff
@@ -0,0 +1,671 @@
+
+    opal_atomic_unlock(&cont_req->cont_lock);
+
+    if (NULL == thread_progress_list) {
```
Why is this global object not protected for concurrent accesses ?

devreal replied:
That's a thread_local variable that stores the continuation requests the current thread is waiting on

devreal committed:
Make sure threads waiting on continuation request execute continuations while waiting

No other threads should execute them if the continuation request is limited
to polling only.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Add tests for MPI continuations (single- and multi-threaded)

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Continuations: updated API and support for error handling

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Continuations: Add callback return value

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Move ompi_mpi_object_t to separate header file

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Error handler: query stored MPI object to invoke error on

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Minor fixes: break include cycle and remove unnecesary atomic

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Atomics: add relaxed load and store

The _Atomic qualifier requires accesses to such variables to
ensure sequential consistency, leading to unnecessary atomic operations.
OPAL_ATOMIC_RELAXED_STORE and OPAL_ATOMIC_RELAXED_LOAD avoid atomic
operations in paths where atomicity and memory ordering is not required.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Use OPAL_ATOMIC_RELAXED_STORE in continuations

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Add explicit start to continuation requests

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Remove locks if request buffer is volatile

If we cannot maintain access to the requests we cannot handle errors.
Thus, there is no need to move the continuations between lists.
Using volatile request buffers in combination with non-aborting error
handlers should be treated as undefined behavior.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Add missing MPIX_Continue_get_failed function implementation

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Don't execute callbacks immediately if we didn't see any completed requests

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix rebase conflict

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix starting and completion of CRs and reduce atomic operations

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Revert "Atomics: add relaxed load and store"

This reverts commit 4518dedc313e95639cc8ebd37879b381d0c0c130.

devreal committed:
Remove use of OPAL_ATOMIC_RELAXED_STORE

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Make sure CRs are restarted in tests

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Don't execute a continuation immediately if the CR is inactive

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Avoid using thread_local data if multi-threading is not enabled

Accessing thread-local data in a shared library is still expensive...

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Use atomic locks instead of pthread locks to avoid library calls

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Make sure the progress callback is always registered

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix logic errors for continuation flags

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Only set in_progress around code that directly executes continuations

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Always return 1 from request_completion_cb

No need to change the req_complete pointer as there are no sync objects
on requests that have a continuation attached.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Reduce atomic operations by checking for empty list

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Reorder progress of individual requests

Explicit global progress instead of global progress
when progressing a request, to avoid recursive progress calls.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Avoid atomic operation if all requests have completed

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Let the test and wait calls set CRs to inactive

Also: always store continuations directly in poll-only CRs.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Progress CR in ompi_request_wait_completion before creating a sync

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Continuation tests: test for delayed start

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Infrastructure to test for MPI object error handler, not used yet

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Remove an atomic op from wait_sync_update

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Outline constructor for thread-local data

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Minor changes to the continuation API to reflect current proposal

- Swap flags and max_poll parameter in MPI_Continue_init
- Swap cont_req and info parameter in MPI_Continue_init
- Add new flags and remove support for info keys mpi_continue_poll_only,
  mpi_continue_max_poll, and mpi_continue_enqueue_complete

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Remove left over use of OPAL_ATOMIC_RELAXED_STORE

This was a temporary bandaid to fix the C11 sequentially consistent store...

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Make sure completed continuations requests are marked pending

Continuation requests may be marked as complete before another
continuation is registered. In that case, we should mark
the continuation request as pending again, unless a sync object
has been registered already.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Don't set persistent requests to MPI_REQUEST_NULL

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix wrong locking when freeing a continuation request

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Continuations: Fix potential race condition in debug mode

The cont->cont_req field was reset *after* it was returned to the freelist...

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Hotfix for lost callbacks

We must protect the activation of the completion of a continuation request
to ensure that we don't miss any updates.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Properly defer continuations if the cont_req is inactive

Runnable continuations should be added to cont_complete_defer_list if
the continuation request is inactive, so that it can be enqueued
for execution once the request is enabled.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Slight code cleanup and remove locks if threads are not enabled

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Docs: don't mention MPI_UNDEFINED

This was a temporary idea that has been removed again.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fixup documentation and remove mention of MPIX_CONT_PERSISTENT

MPIX_CONT_PERSISTENT will not be part of the first proposal and has
been removed.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Continuations: Cleanup dox and implement invoke-failed

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Put a continuation back into the cont_incomplete_list if rechecking

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Add write memory barrier before invoking request callback

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Add REQUEST_CB_PENDING that got lost during rebase

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Fix compiler errors in continuation.c

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
Don't cast function pointers to void*

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
pml/ucx: Properly reinitialize requests

It is not sufficient to just set one value, we need proper initialization
through OMPI_REQUEST_INIT.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

JiakunYan commented:
I found something that may need more clarification.

The continuation proposal says

> A continuation request completes once the callbacks of all continuations registered with it have completed execution.
After initialization or completion, a continuation request has to be started to enable the execution of registered continuations.

It seems to me that the MPI runtime should set the status of the continuation request to `INACTIVE` when the last continuation registered with it completes its execution. However, what I see in this PR is that the code will set the status to `INACTIVE` only when `MPI_Test` on the continuation request finds it completed.

This difference matters because it decides whether the following code logic is correct: if my application never tested (and then re-started) the continuation request (imagining I have some magic way to keep making progress on the runtime e.g., testing other regular requests), could I keep attaching new continues to it and assume they always get executed without delay?

devreal commented:
I think this is a good point. The idea was that CRs *may* complete once all registered continuations have executed and the CR is tested/waited on. It will not be complete otherwise, so it is safe to register new continuations and expect them to be executed (provided they are not poll-only).

hppritcha commented:
@devreal are you still working on this?

devreal commented:
I am working on it but we can close the PR. I continue working on the branch and bring it back once complete.
